lookout/style/format/rules.py

```diff
@@ -31,17 +31,110 @@
+        super().__init__(**kwargs)
+        self._rules_by_lang = {}
+
+    @property
+    def languages(self):
+        return sorted(self._rules_by_lang)
+
+    def construct(self, ruless: Iterable[Tuple[str, "Rules"]]) -> "FormatModel":
+        for name, rules in ruless:
+            self[name] = rules
+        return self
 
     def dump(self) -> str:
-        pass
+        if len(self) == 0:
+            return "<empty FormatModel>"
+        languages = self.languages
+        param_names = self[languages[0]]._get_param_names()
+        return "Model languages: %s.\n" \
+               "First model's params: %s\n" \
+               "First model's rules number: %d.\n" % \
+               (str(languages),
+                ", ".join(["%s=%s" % (p, str(getattr(self[languages[0]], p)))
+                           for p in param_names]),
+                len(self[languages[0]]._rules))
 
     def _generate_tree(self) -> dict:
-        pass
+        languages = self.languages
+        return dict(
+            languages=languages,
+            paramss=[self[lang].get_params(deep=False) for lang in languages],
+            ruless=[self._disassemble_rules(self[lang]._rules) for lang in languages],
+        )
 
     def _load_tree(self, tree: dict) -> None:
-        pass
+        for name, params, rules in zip(tree["languages"], tree["paramss"], tree["ruless"]):
+            params = dict(params)
+            params["top_down_greedy_budget"] = Rules.TopDownGreedyBudget(
+                *params["top_down_greedy_budget"])
+            params["_rules"] = self._assemble_rules(rules)
+            _rules = Rules.__new__(Rules)
+            _rules.__setstate__(params)
+            self[name] = _rules
+
+    def __len__(self) -> int:
+        return len(self._rules_by_lang)
+
+    def __getitem__(self, lang: str) -> "Rules":
+        """
+        Get the Rules estimator by its language.
+        :param lang: Estimator language.
+        :return: Rules estimator instance.
+        """
+        return self._rules_by_lang[lang]
+
+    def __setitem__(self, lang: str, rules: "Rules"):
+        """
+        Set a new Rules estimator to the model by its language.
+        """
+        if not rules.fitted:
+            raise NotFittedError("Rules estimator should be fitted before adding to FormatModel.")
+        self._rules_by_lang[lang] = rules
+
+    def __iter__(self):
+        yield from self._rules_by_lang.__iter__()
+
+    def __contains__(self, item):
+        return item in self._rules_by_lang
+
+    @staticmethod
+    def _assemble_rules(rules_tree: dict) -> List[Rule]:
+        rules = []
+        rule_attrs = (RuleAttribute(*params) for params in
+                      zip(rules_tree["features"],  rules_tree["cmps"], rules_tree["thresholds"]))
+        for cls, conf, length in zip(rules_tree["cls"], rules_tree["conf"], rules_tree["lengths"]):
+            rules.append(Rule(tuple(islice(rule_attrs, length)), RuleStats(cls, conf)))
```

Comment:
Tests fail here. I guess because `length` can appear a numpy.int here which is not a subclass of int. So you should ensure that you cast to `int` in `assemble()`